src/meta/src/barrier/rpc.rs

```diff
@@ -109,17 +110,33 @@ impl ControlStreamManager {
             .committed_epoch;
         let node_id = node.id;
         let node_host = node.host.clone().unwrap();
-        match self.context.new_control_stream_node(node, prev_epoch).await {
-            Ok((stream_node, response_stream)) => {
-                let _ = self.nodes.insert(node_id, stream_node);
-                self.response_streams
-                    .push(into_future(node_id, response_stream));
-                info!(?node_host, "add control stream worker");
-            }
-            Err(e) => {
-                error!(err = %e.as_report(), ?node_host, "fail to start control stream with worker node");
+        let mut backoff = ExponentialBackoff::from_millis(100)
+            .max_delay(Duration::from_secs(3))
+            .factor(5);
+        const MAX_RETRY: usize = 5;
+        for i in 1..=MAX_RETRY {
+            match self
+                .context
+                .new_control_stream_node(node.clone(), prev_epoch)
+                .await
+            {
+                Ok((stream_node, response_stream)) => {
+                    let _ = self.nodes.insert(node_id, stream_node);
+                    self.response_streams
+                        .push(into_future(node_id, response_stream));
+                    info!(?node_host, "add control stream worker");
+                    return;
+                }
+                Err(e) => {
+                    // It may happen that the dns information of newly registered worker node
+                    // has not been propagated to the meta node and cause error. Wait for a while and retry
+                    let delay = backoff.next().unwrap();
+                    error!(attempt = i, backoff_delay = ?delay, err = %e.as_report(), ?node_host, "fail to resolve worker node address");
+                    sleep(delay).await;
+                }
             }
         }
+        error!(?node_host, "fail to create worker node after retry");
```

Comment:
Should we retry indefinitely or mark the node as inactive or panic instead of just printing an error log if we cannot create the streaming RPC? 

With the previous behavior using unary RPC, if meta cannot connect to a worker node, I think InjectBarrier RPC will fail, leading to recovery. The recovery process will continue until InjectBarrier RPC can succeed, which is equivalent to retry indefinitely.

Reply:
If the stream cannot be created, it will not exist in the node map, and then injecting barrier will return with error because the worker node to be injected can not be found. Actually this is the cause of the previous failure to create table.

Not adding the worker node to the node map is something like marking the node as inactive.

Reply:
> With the previous behavior using unary RPC, if meta cannot connect to a worker node, I think InjectBarrier RPC will fail, leading to recovery.

Correction: InjectBarrier RPC will not be sent to worker node without any actors so the recovery and retry will only happen when new actors are created or actor rebalance happens on the unreachable worker node. But still with this PR, as long as the initial `add_worker` attempt fails, there won't be future retry even when new actors are assigned to the worker node.

Reply:
> > With the previous behavior using unary RPC, if meta cannot connect to a worker node, I think InjectBarrier RPC will fail, leading to recovery.
> 
> Correction: InjectBarrier RPC will not be sent to worker node without any actors so the recovery and retry will only happen when new actors are created or actor rebalance happens on the unreachable worker node. But still with this PR, as long as the initial `add_worker` attempt fails, there won't be future retry even when new actors are assigned to the worker node.

If a CN does not have any actor on it, failure to connect to it should not trigger recovery. When new actors are assigned to the worker node, there will be a inject barrier request to the CN, and then it will return an error when unable to find the CN in the node map, and then trigger the recovery, and then retry will be handled in the recovery loop.

Since this is a blocking async method, endless retrying in the method may block the usual handling of events.

Reply:
> When new actors are assigned to the worker node, there will be a inject barrier request to the CN, and then it will return an error when unable to find the CN in the node map, and then trigger the recovery, and then retry will be handled in the recovery loop.

IIUC, if the retry attempts are exhausted in the initial `add_worker` call, the node map won't have an entry for the worker node. In this case, when there are new actors assigned to the worker node and the inject barrier fails, will recovery trigger `add_worker` again and try to re-establish the streaming RPC?

Reply:
Oh my bad. recovery will trigger `reset`, which will try re-establish the streaming RPC. Then why did the tests fail prior to this PR? Does it mean that the failing tests expect no recovery triggered at all, even though things can be resolve after several rounds of recovery.

Reply:
> Then why did the tests fail prior to this PR?

In the longevity test, there is an initial SQL to create table. The create table SQL fails without retry, and then all subsequent create MV SQL failed.